common/cuda_hip/reorder/rcm_kernels.hpp.inc

```diff
@@ -0,0 +1,614 @@
+{
+    const auto policy = thrust_policy(exec);
+    components_data<IndexType> result{exec, static_cast<size_type>(num_rows)};
+    const auto node_component = result.node_component.get_data();
+    const auto nodes = result.nodes.get_data();
+    // attach every node to a smaller neighbor
+    const auto num_blocks = ceildiv(num_rows, default_block_size);
+    connected_components_attach<<<num_blocks, default_block_size, 0,
+                                  exec->get_stream()>>>(
+        row_ptrs, col_idxs, num_rows, node_component);
+    // combine connected components along edges
+    connected_components_combine<<<num_blocks, default_block_size, 0,
+                                   exec->get_stream()>>>(
+        row_ptrs, col_idxs, num_rows, node_component);
+    // compress paths to edges
+    connected_components_path_compress<<<num_blocks, default_block_size, 0,
+                                         exec->get_stream()>>>(num_rows,
+                                                               node_component);
+    // group nodes by component ID
+    result.sorted_ids = result.node_component;
+    const auto sorted_component_ids = result.sorted_ids.get_data();
+    thrust::sequence(policy, nodes, nodes + num_rows, IndexType{});
+    thrust::stable_sort_by_key(policy, sorted_component_ids,
+                               sorted_component_ids + num_rows, nodes);
+    // find beginning of all components
+    auto it = thrust::make_counting_iterator(size_type{});
+    const auto predicate = adj_not_predicate<IndexType>{sorted_component_ids};
+    const auto num_components = static_cast<size_type>(
+        thrust::count_if(policy, it, it + num_rows, predicate));
+    result.set_num_components(num_components);
+    const auto ptrs = result.ptrs.get_data();
+    const auto min_deg_node = result.min_deg_node.get_data();
+    thrust::copy_if(policy, it, it + num_rows, ptrs, predicate);
+    // set the sentinel entry
+    set_element(result.ptrs, num_components, num_rows);
+    // find minimum degree node for each component
+    array<IndexType> component_id_array{exec, num_components};
+    const auto component_ids = component_id_array.get_data();
+    thrust::reduce_by_key(
+        policy, sorted_component_ids, sorted_component_ids + num_rows, nodes,
+        component_ids, min_deg_node, thrust::equal_to<IndexType>{},
+        node_min_degree_reduction<IndexType>{num_rows, node_degrees});
+    // map component IDs to consecutive indexing
+    array<IndexType> compacted_node_component{exec,
+                                              static_cast<size_type>(num_rows)};
+    thrust::lower_bound(policy, component_ids, component_ids + num_components,
+                        node_component, node_component + num_rows,
+                        compacted_node_component.get_data());
+    result.node_component = std::move(compacted_node_component);
+    return result;
+}
+
+
+template <typename IndexType>
+struct ubfs_levels {
```

Comment:
I know bfs should be bradth-first search, but what's the meaning of `u` here?

Reply:
unordered, since we use atomics to write out the next level and sort afterwards